sw-emulator/lib/periph/src/abr.rs

```diff
@@ -0,0 +1,2347 @@
+
+        self.kv_mlkem_seed_rd_status.reg.modify(
+            KvRdSeedStatus::READY::SET
+                + KvRdSeedStatus::VALID::SET
+                + KvRdSeedStatus::ERROR.val(seed_read_result),
+        );
+    }
+
+    fn mlkem_msg_read_complete(&mut self) {
+        let key_id = self.kv_mlkem_msg_rd_ctrl.reg.read(KvRdSeedCtrl::READ_ENTRY);
+        let msg_read_result = self.mlkem_read_msg_from_keyvault(key_id, false);
+
+        self.kv_mlkem_msg_rd_status.reg.modify(
+            KvRdSeedStatus::READY::SET
+                + KvRdSeedStatus::VALID::SET
+                + KvRdSeedStatus::ERROR.val(msg_read_result),
+        );
+    }
+
+    fn mlkem_sharedkey_write_complete(&mut self) {
+        let key_id = self
+            .kv_mlkem_sharedkey_wr_ctrl
+            .reg
+            .read(KeyWriteControl::KEY_ID);
+        let key_usage = self
+            .kv_mlkem_sharedkey_wr_ctrl
+            .reg
+            .read(KeyWriteControl::USAGE);
+        let write_result = self.mlkem_write_sharedkey_to_keyvault(key_id, key_usage);
+
+        self.kv_mlkem_sharedkey_wr_status.reg.modify(
+            KvRdSeedStatus::READY::SET
+                + KvRdSeedStatus::VALID::SET
+                + KvRdSeedStatus::ERROR.val(write_result),
+        );
+    }
+
+    fn mlkem_gen_key(&mut self) {
+        let seed_d = bytes_from_words_le(&self.mlkem_seed_d);
+        let seed_z = bytes_from_words_le(&self.mlkem_seed_z);
+
+        // Use both seeds for ML-KEM keygen
+        let mut rng = SeedOnlyRng::new_with_seeds(vec![seed_d, seed_z]);
+        let (dk, ek) = MlKem1024::generate(&mut rng);
+
+        // Copy to register arrays
+        let dk_bytes = dk.as_bytes();
+        for (i, chunk) in dk_bytes.chunks(4).enumerate() {
+            if i < self.mlkem_decaps_key.len() {
+                let mut word = [0u8; 4];
+                word[..chunk.len()].copy_from_slice(chunk);
+                self.mlkem_decaps_key[i] = u32::from_le_bytes(word);
+            }
+        }
+
+        let ek_bytes = ek.as_bytes();
+        for (i, chunk) in ek_bytes.chunks(4).enumerate() {
+            if i < self.mlkem_encaps_key.len() {
+                let mut word = [0u8; 4];
+                word[..chunk.len()].copy_from_slice(chunk);
+                self.mlkem_encaps_key[i] = u32::from_le_bytes(word);
+            }
+        }
+    }
+
+    fn mlkem_encaps(&mut self) {
+        // Reconstruct encapsulation key from register
+        let ek_bytes = bytes_from_words_le(&self.mlkem_encaps_key);
+        let ek = EncapsulationKey::<MlKem1024Params>::from_bytes(
+            ek_bytes.as_slice().try_into().unwrap(),
+        );
+
+        let message = bytes_from_words_le(&self.mlkem_msg);
+        let mut rng = SeedOnlyRng::new(message);
+
+        match ek.encapsulate(&mut rng) {
+            Ok((ct, ss)) => {
+                // Store shared secret
+                self.mlkem_shared_key =
```

Comment:
The shared key and decapsulation keys should not be visible if the seed came from KV.